src/client/Client.cc

```diff
@@ -15944,8 +15944,10 @@ int64_t Client::ll_preadv_pwritev(struct Fh *fh, const struct iovec *iov,
                                   bool do_fsync, bool syncdataonly)
 {
     RWRef_t mref_reader(mount_state, CLIENT_MOUNTING);
-    if (!mref_reader.is_state_satisfied())
-      return -CEPHFS_ENOTCONN;
```

Comment:
@ffilz the nullptr when provided to the async api leads to a crash https://tracker.ceph.com/issues/63633, we had a discussion about this on the upstream cephfs slack channel https://ceph-storage.slack.com/archives/C04LVQMHM9B/p1700743336859379 where we concluded that since the async api doesn't work with nullptr context (i.e. it must've been designed like this since an async call without context makes no sense); we must assert that `context != null` __but__ when you mentioned that a nullptr context should make it a sync call(i had no idea about this behaviour before) in https://github.com/ceph/ceph/pull/54651#discussion_r1414306203, it made it clear that the api is indeed buggy and needs to be fixed. There's this PR https://github.com/ceph/ceph/pull/54808, for now this has just the test case but soon i will come up with the RCA and a fix. If you have any insights on this, do let me know.

Reply:
One option is to declare ll_preadv_pwritev async only, so remove the onfinish = nullptr from the declaration in client.h. So far it IS only used for async. But ll_readv and ll_writev both call _preadv_pwritev_locked with null onfinish, so unless they are also broken, there should not be anything wrong with calling ll_preadv_pwritev with onfinish null.

It looks like the crash was in a particular test case, what was the ll_preadv_pwritev invocation that caused a crash?

I suspect some combination of parameters with onfinish null is causing a problem

Reply:
PR https://github.com/ceph/ceph/pull/54808 ready for review

Reply:
with this i cannot directly complete the context directly, demands checking for context and act accordingly just like the main commit of this PR https://github.com/ceph/ceph/pull/54690/commits/713b3fc5c3eff6216a8f3f0c9bccf7899f7c463a and this brings us back to starting point i.e. to have a function like `maybe_finish_context()` since having similar checks twice within same code patch doesn't make sense